File: tests/fixtures/xpath_injection/php/unsafe_xpath_query.php
<?php
// Unsafe: $_GET['user'] concatenated into an XPath expression and passed
// straight to SimpleXMLElement::xpath.  XPATH_INJECTION fires on the
// expression argument.
$xml = simplexml_load_file("users.xml");
$user = $_GET['user'];
$expr = "//user[name='" . $user . "']";
$nodes = $xml->xpath($expr);

File: tests/fixtures/xpath_injection/java/UnsafeXPathQuery.java
// Unsafe: attacker-controlled username concatenated into an XPath expression
// passed to XPath.evaluate.  The flat matcher catches the qualified call.
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathFactory;
import javax.servlet.http.HttpServletRequest;
import org.w3c.dom.Document;
import org.w3c.dom.NodeList;

public class UnsafeXPathQuery {
    public NodeList lookup(HttpServletRequest req, Document doc) throws Exception {
        String user = req.getParameter("user");
        String expr = "//user[name='" + user + "']";
        XPath xpath = XPathFactory.newInstance().newXPath();
        return (NodeList) xpath.evaluate(expr, doc, XPathConstants.NODESET);
    }
}
